include/nbl/builtin/glsl/sampling/spherical_rectangle.glsl

```diff
@@ -0,0 +1,34 @@
+// Copyright (C) 2018-2022 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_BUILTIN_GLSL_SAMPLING_SPHERICAL_RECTANGLE_INCLUDED_
+#define _NBL_BUILTIN_GLSL_SAMPLING_SPHERICAL_RECTANGLE_INCLUDED_
+
+#include <nbl/builtin/glsl/math/quaternions.glsl>
+#include <nbl/builtin/glsl/shapes/rectangle.glsl>
+
+#define CLAMP_EPS 1e-5f
+
+// Code from https://www.arnoldrenderer.com/research/egsr2013_spherical_rectangle.pdf
+vec3 nbl_glsl_sampling_generateSphericalRectangleSample(in vec3 r0, in vec3 r1, in float S, in float b0, float b1, in float k, vec2 uv)
+{
+    // 1. compute ’cu’
+    float au = uv.x * S + k;
+    float fu = (cos(au) * b0 - b1) / sin(au);
+    float cu = 1/sqrt(fu*fu + b0 * b0) * (fu>0 ? +1 : -1);
+    cu = clamp(cu, -1, 1); // avoid NaNs
+    // 2. compute ’xu’
+    float xu = -(cu * r0.z) / sqrt(1 - cu*cu);
+    xu = clamp(xu, r0.x, r1.x); // avoid Infs
+    // 3. compute ’yv’
+    float d = sqrt(xu*xu + r0.z*r0.z);
```

Comment:
```glsl
const float cu_2 = max(fu*fu+b0_2,1.f); // forces `cu` to be in [-1,1]
const float cu = uintBitsToFloat(floatBitsToUint(inversesqrt(cu_2))^(floatBitsToUint(fu)&0x80000000u));

const float one_minus_cu_2 = 1.f-cu_2;
// would have used `abs(r0.z)` but r0.z is guaranteed not to be positive
const float d = -r0.z*inversesqrt(one_minus_cu_2);
const float xu = d<=nbl_glsl_FLT_MAX (-cu*d):r0.x; // avoid Infs

const float d_2 = r0_2.z/one_minus_cu_2;
```

Reply:
seems to have a bug, re-derive

Reply:
Done, but feel free to tell if it's done wrong.